<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

# Working with a real world data-set using SQL and Python

Estaimted time needed: **30** minutes

## Objectives

After complting this lab you will be able to:

*   Understand the dataset for Chicago Public School level performance
*   Store the dataset in SQLite database.
*   Retrieve metadata about tables and columns and query data from mixed case columns
*   Solve example problems to practice your SQL skills including using built-in database functions


## Chicago Public Schools - Progress Report Cards (2011-2012)

The city of Chicago released a dataset showing all school level performance data used to create School Report Cards for the 2011-2012 school year. The dataset is available from the Chicago Data Portal: [https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

This dataset includes a large number of metrics. Start by familiarizing yourself with the types of metrics in the database: [https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?download=true](https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&download=true&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

**NOTE**:

Do not download the dataset directly from City of Chicago portal. Instead download a static copy which is a more database friendly version from this <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">link</a>.

Now review some of its contents.


### Connect to the database

Let us now load the ipython-sql  extension and establish a connection with the database

##### The syntax for connecting to magic sql using sqllite is

  **%sql sqlite://DatabaseName**   
  
where DatabaseName will be your **.db** file 


In [1]:
import csv, sqlite3

con = sqlite3.connect("RealWorldData.db")
cur = con.cursor()

In [2]:
!pip install ibm_db_sa
!pip install ipython-sql
!pip install -q pandas==1.1.5

'pip' is not recognized as an internal or external command,
operable program or batch file.
'pip' is not recognized as an internal or external command,
operable program or batch file.
'pip' is not recognized as an internal or external command,
operable program or batch file.


In [3]:
%load_ext sql

In [4]:
%sql sqlite:///RealWorldData.db

### Store the dataset in a Table

##### In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. To analyze the data 
##### using SQL, it first needs to be stored in the database.

##### We will first read the csv files  from the given url  into pandas dataframes

##### Next we will be using the  df.to_sql() function to convert each csv file  to a table in sqlite  with the csv data loaded in it.


In [5]:
import pandas as pd
chicago_school_df =pd.read_csv("ChicagoPublicSchools.csv", sep=",")
chicago_school_df.to_sql("chicago_school", con, if_exists='replace', index=False)
chicago_school_df.head()

C:\Users\Mila\AppData\Roaming\Python\Python311\site-packages\pandas\core\generic.py:2872: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.
  sql.to_sql(


School_ID                                     NAME_OF_SCHOOL  \
0     610038                  Abraham Lincoln Elementary School   
1     610281  Adam Clayton Powell Paideia Community Academy ...   
2     610185                Adlai E Stevenson Elementary School   
3     609993                    Agustin Lara Elementary Academy   
4     610513                      Air Force Academy High School   

  Elementary, Middle, or High School         Street_Address     City State  \
0                                 ES        615 W Kemper Pl  Chicago    IL   
1                                 ES  7511 S South Shore Dr  Chicago    IL   
2                                 ES     8010 S Kostner Ave  Chicago    IL   
3                                 ES     4619 S Wolcott Ave  Chicago    IL   
4                                 HS        3630 S Wells St  Chicago    IL   

   ZIP_Code    Phone_Number  \
0     60614  (773) 534-5720   
1     60649  (773) 535-6650   
2     60652  (773) 535-2280   
3     60609  (773) 535-4389   
4     60609  (773) 535-1590   

                                                Link  \
0  http://schoolreports.cps.edu/SchoolProgressRep...   
1  http://schoolreports.cps.edu/SchoolProgressRep...   
2  http://schoolreports.cps.edu/SchoolProgressRep...   
3  http://schoolreports.cps.edu/SchoolProgressRep...   
4  http://schoolreports.cps.edu/SchoolProgressRep...   

                      Network_Manager  ... Freshman_on_Track_Rate__  \
0        Fullerton Elementary Network  ...                      NDA   
1           Skyway Elementary Network  ...                      NDA   
2           Midway Elementary Network  ...                      NDA   
3         Pershing Elementary Network  ...                      NDA   
4  Southwest Side High School Network  ...                     91.8   

  X_COORDINATE Y_COORDINATE   Latitude  Longitude COMMUNITY_AREA_NUMBER  \
0  1171699.458  1915829.428  41.924497 -87.644522                     7   
1  1196129.985  1856209.466  41.760324 -87.556736                    43   
2  1148427.165  1851012.215  41.747111 -87.731702                    70   
3  1164504.290  1873959.199  41.809757 -87.672145                    61   
4  1175177.622  1880745.126  41.828146 -87.632794                    34   

  COMMUNITY_AREA_NAME  Ward Police_District                     Location  
0        LINCOLN PARK    43              18  (41.92449696, -87.64452163)  
1         SOUTH SHORE     7               4  (41.76032435, -87.55673627)  
2             ASHBURN    13               8  (41.74711093, -87.73170248)  
3            NEW CITY    20               9    (41.8097569, -87.6721446)  
4       ARMOUR SQUARE    11               9  (41.82814609, -87.63279369)  

[5 rows x 78 columns]

Double-click **here** for the solution.
<!--

import pandas
df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv")
df.to_sql("CENSUS_DATA", con, if_exists='replace', index=False,method="multi")

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv")
df.to_sql("CHICAGO_CRIME_DATA", con, if_exists='replace', index=False, method="multi")

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv")
df.to_sql("CHICAGO_PUBLIC_SCHOOLS_DATA", con, if_exists='replace', index=False, method="multi")

-->


### Query the database system catalog to retrieve table metadata

##### You can verify that the table creation was successful by retrieving the list of all tables in your schema and checking whether the SCHOOLS table was created


In [6]:
# type in your query to retrieve list of all tables in the database
cur.execute(f"PRAGMA table_info(chicago_school);")
info = cur.fetchall()

if len(info):
    print("Table exists")
else:
    print("Table does not exist")

Table exists


Double-click **here** for a hint

<!--
In sqlite the system catalog table called sqlite_master contains the table metadata
-->


Double-click **here** for the solution.

<!-- Solution using SQLite:

%sql SELECT name FROM sqlite_master WHERE type='table'

-->



### Query the database system catalog to retrieve column metadata

##### The SCHOOLS table contains a large number of columns. How many columns does this table have?


In [7]:
# type in your query to retrieve the number of columns in the SCHOOLS table
cur.execute(f"PRAGMA table_info(chicago_school);")
info = cur.fetchall()
print(len(info))


78


Double-click **here** for the solution.

<!-- Solution SQLite:
%sql SELECT count(name) FROM PRAGMA_TABLE_INFO('CHICAGO_PUBLIC_SCHOOLS_DATA');

-->




Now retrieve the the list of columns in SCHOOLS table and their column type (datatype) and length.


In [8]:
cur.execute(f"PRAGMA table_info(chicago_school);")
info = cur.fetchall()
df=pd.DataFrame(info, columns=["ID", "Name", "Type", "NotNull", "Default", "Primary Key"])
df.head()

ID                                Name     Type  NotNull Default  \
0   0                           School_ID  INTEGER        0    None   
1   1                      NAME_OF_SCHOOL     TEXT        0    None   
2   2  Elementary, Middle, or High School     TEXT        0    None   
3   3                      Street_Address     TEXT        0    None   
4   4                                City     TEXT        0    None   

   Primary Key  
0            0  
1            0  
2            0  
3            0  
4            0

Double-click **here** for the solution.

<!-- Solution 

%sql SELECT name,type,length(type) FROM PRAGMA_TABLE_INFO('CHICAGO_PUBLIC_SCHOOLS_DATA');

-->




### Questions

1.  Is the column name for the "SCHOOL ID" attribute in upper or mixed case?
2.  What is the name of "Community Area Name" column in your table? Does it have spaces?
3.  Are there any columns in whose names the spaces and paranthesis (round brackets) have been replaced by the underscore character "\_"?


## Problems

### Problem 1

##### How many Elementary Schools are in the dataset?


In [9]:
statment='''SELECT COUNT(*) FROM chicago_school WHERE "Elementary, Middle, or High School"='ES';'''
cur.execute(statment)
rows = cur.fetchall()
print(rows[0][0])


462


Double-click **here** for a hint

<!--
Which column specifies the school type e.g. 'ES', 'MS', 'HS'? ("Elementary School, Middle School, High School")
-->


Double-click **here** for another hint

<!--
Does the column name have mixed case, spaces or other special characters?
If so, ensure you use double quotes around the "Name of the Column"
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select count(*) from CHICAGO_PUBLIC_SCHOOLS_DATA where "Elementary, Middle, or High School"='ES'

Correct answer: 462

-->


### Problem 2

##### What is the highest Safety Score?


In [10]:
statment='''SELECT MAX(SAFETY_SCORE) FROM chicago_school'''
cur.execute(statment)
rows = cur.fetchone()
print(rows[0])

99.0


Double-click **here** for a hint

<!--
Use the MAX() function
-->


Double-click **here** for the solution.

<!-- Hint:

%sql select MAX(Safety_Score) AS MAX_SAFETY_SCORE from CHICAGO_PUBLIC_SCHOOLS_DATA

Correct answer: 99
-->


### Problem 3

##### Which schools have highest Safety Score?


In [11]:
statment='''SELECT NAME_OF_SCHOOL, MAX(SAFETY_SCORE) as safety FROM chicago_school;'''
cur.execute(statment)
rows = cur.fetchall()
print(rows[0][0])


Abraham Lincoln Elementary School


Double-click **here** for the solution.

<!-- Solution:
In the previous problem we found out that the highest Safety Score is 99, so we can use that as an input in the where clause:

%sql select Name_of_School, Safety_Score from CHICAGO_PUBLIC_SCHOOLS_DATA where Safety_Score = 99

or, a better way:

%sql select Name_of_School, Safety_Score from CHICAGO_PUBLIC_SCHOOLS_DATA where \
  Safety_Score= (select MAX(Safety_Score) from CHICAGO_PUBLIC_SCHOOLS_DATA)


Correct answer: several schools with with Safety Score of 99.
-->


### Problem 4

##### What are the top 10 schools with the highest "Average Student Attendance"?


In [12]:
statment='''SELECT MAX(AVERAGE_STUDENT_ATTENDANCE) FROM chicago_school LIMIT 10;'''
cur.execute(statment)
rows = cur.fetchone()
print(rows[0])

98.40%


Double-click **here** for the solution.

<!-- Solution:

%sql select Name_of_School, Average_Student_Attendance from CHICAGO_PUBLIC_SCHOOLS_DATA \
    order by Average_Student_Attendance desc nulls last limit 10 

-->


### Problem 5

##### Retrieve the list of 5 Schools with the lowest Average Student Attendance sorted in ascending order based on attendance


In [14]:
statement='''select NAME_OF_SCHOOL, AVERAGE_STUDENT_ATTENDANCE from chicago_school order by AVERAGE_STUDENT_ATTENDANCE limit 5;'''
cur.execute(statement)
rows = cur.fetchall()
df=pd.DataFrame(rows, columns=["NAME_OF_SCHOOL", "AVERAGE_STUDENT_ATTENDANCE"])
df.head()


NAME_OF_SCHOOL  \
0              Velma F Thomas Early Childhood Center   
1  Richard T Crane Technical Preparatory High School   
2       Barbara Vick Early Childhood & Family Center   
3                                  Dyett High School   
4               Wendell Phillips Academy High School   

  AVERAGE_STUDENT_ATTENDANCE  
0                       None  
1                     57.90%  
2                     60.90%  
3                     62.50%  
4                     63.00%

Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     order by Average_Student_Attendance \
     LIMIT 5

-->


### Problem 6

##### Now remove the '%' sign from the above result set for Average Student Attendance column


In [16]:

df['AVERAGE_STUDENT_ATTENDANCE'] = df['AVERAGE_STUDENT_ATTENDANCE'].replace('%', '', regex=True)
df.head()

NAME_OF_SCHOOL  \
0              Velma F Thomas Early Childhood Center   
1  Richard T Crane Technical Preparatory High School   
2       Barbara Vick Early Childhood & Family Center   
3                                  Dyett High School   
4               Wendell Phillips Academy High School   

   AVERAGE_STUDENT_ATTENDANCE  
0                         NaN  
1                        57.9  
2                        60.9  
3                        62.5  
4                        63.0

Double-click **here** for a hint

<!--
Use the REPLACE() function to replace '%' with ''

-->


Double-click **here** for the solution.

<!-- Hint:

%sql SELECT Name_of_School, REPLACE(Average_Student_Attendance, '%', '') \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     order by Average_Student_Attendance \
     LIMIT 5

-->


### Problem 7

##### Which Schools have Average Student Attendance lower than 70%?


In [19]:
statement='''SELECT NAME_OF_SCHOOL,AVERAGE_STUDENT_ATTENDANCE FROM chicago_school WHERE AVERAGE_STUDENT_ATTENDANCE < "70%"'''
cur.execute(statement)
rows = cur.fetchall()
df=pd.DataFrame(rows, columns=["NAME_OF_SCHOOL", "AVERAGE_STUDENT_ATTENDANCE"])
df.head()

NAME_OF_SCHOOL AVERAGE_STUDENT_ATTENDANCE
0   Barbara Vick Early Childhood & Family Center                     60.90%
1  Chicago Vocational Career Academy High School                     68.80%
2                              Dyett High School                     62.50%
3              Manley Career Academy High School                     66.80%
4                        Orr Academy High School                     66.30%

Double-click **here** for a hint

<!--
The datatype of the "Average_Student_Attendance" column is varchar.
So you cannot use it as is in the where clause for a numeric comparison.
First use the CAST() function to cast it as a DECIMAL or DOUBLE
e.g. CAST("Column_Name" as DOUBLE)

-->


Double-click **here** for another hint

<!--
Don't forget the '%' age sign needs to be removed before casting
-->


Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     where CAST ( REPLACE(Average_Student_Attendance, '%', '') AS DOUBLE ) < 70 \
     order by Average_Student_Attendance
     

-->


### Problem 8

##### Get the total College Enrollment for each Community Area


In [20]:
statement='''SELECT sum(COLLEGE_ENROLLMENT),COMMUNITY_AREA_NAME FROM chicago_school GROUP BY COLLEGE_ENROLLMENT'''
cur.execute(statement)
rows = cur.fetchall()
df=pd.DataFrame(rows, columns=["sum(COLLEGE_ENROLLMENT)", "COMMUNITY_AREA_NAME"])
df

sum(COLLEGE_ENROLLMENT) COMMUNITY_AREA_NAME
0                         21      NEAR WEST SIDE
1                         26     CALUMET HEIGHTS
2                         44           ENGLEWOOD
3                         48      NORTH LAWNDALE
4                         83      NORTH LAWNDALE
..                       ...                 ...
429                     2366         IRVING PARK
430                     2883       BRIGHTON PARK
431                     2922        NORWOOD PARK
432                     3320      ARCHER HEIGHTS
433                     4368        NORTH CENTER

[434 rows x 2 columns]

Double-click **here** for a hint

<!--
Verify the exact name of the Enrollment column in the database
Use the SUM() function to add up the Enrollments for each Community Area
-->


Double-click **here** for another hint

<!--
Don't forget to group by the Community Area
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from CHICAGO_PUBLIC_SCHOOLS_DATA \
   group by Community_Area_Name 

-->


### Problem 9

##### Get the 5 Community Areas with the least total College Enrollment  sorted in ascending order


In [21]:
statement='''SELECT sum(COLLEGE_ENROLLMENT) as enrollment ,COMMUNITY_AREA_NAME FROM chicago_school GROUP BY COLLEGE_ENROLLMENT  ORDER BY enrollment LIMIT 5'''
cur.execute(statement)
rows = cur.fetchall()
df=pd.DataFrame(rows, columns=["enrollment", "COMMUNITY_AREA_NAME"])
df

enrollment COMMUNITY_AREA_NAME
0          21      NEAR WEST SIDE
1          26     CALUMET HEIGHTS
2          44           ENGLEWOOD
3          48      NORTH LAWNDALE
4          83      NORTH LAWNDALE

Double-click **here** for a hint

<!--
Order the previous query and limit the number of rows you fetch
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from CHICAGO_PUBLIC_SCHOOLS_DATA \
   group by Community_Area_Name \
   order by TOTAL_ENROLLMENT asc \
   LIMIT 5 

-->


### Problem 10

##### List 5 schools with lowest safety score.


In [23]:
statment='''SELECT NAME_OF_SCHOOL ,SAFETY_SCORE FROM chicago_school where safety_score !='None' ORDER BY SAFETY_SCORE LIMIT 5'''
cur.execute(statment)
rows = cur.fetchall()
df=pd.DataFrame(rows, columns=["NAME_OF_SCHOOL", "SAFETY_SCORE"])
df

NAME_OF_SCHOOL  SAFETY_SCORE
0          Edmond Burke Elementary School           1.0
1          Luke O'Toole Elementary School           5.0
2       George W Tilton Elementary School           6.0
3           Foster Park Elementary School          11.0
4  Emil G Hirsch Metropolitan High School          13.0

Double-click **here** for the solution.

<!-- Solution SQLite:

%sql SELECT name_of_school, safety_score \
FROM CHICAGO_PUBLIC_SCHOOLS_DATA  where safety_score !='None' \
ORDER BY safety_score \
LIMIT 5

-->




### Problem 11

##### Get the hardship index for the community area which has College Enrollment of 4368


Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%%sql 
select hardship_index from CENSUS_DATA CD, CHICAGO_PUBLIC_SCHOOLS_DATA CPS 
where CD.community_area_number = CPS.community_area_number 
and college_enrollment = 4368


-->


### Problem 12

##### Get the hardship index for the community area which has the highest value for College Enrollment


Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%sql select community_area_number, community_area_name, hardship_index from CENSUS_DATA \
   where community_area_number in \
   ( select community_area_number from CHICAGO_PUBLIC_SCHOOLS_DATA order by college_enrollment desc limit 1 )
-->


## Summary

##### In this lab you learned how to work with a real word dataset using SQL and Python. You learned how to query columns with spaces or special characters in their names and with mixed case names. You also used built in database functions and practiced how to sort, limit, and order result sets, as well as used sub-queries and worked with multiple tables.


## Author

<a href="https://www.linkedin.com/in/ravahuja/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Rav Ahuja</a>

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By        | Change Description                        |
| ----------------- | ------- | ----------------- | ----------------------------------------- |
| 2022-03-04        | 2.2     | Lakshmi Holla     | Made changes in markdown cells |
| 2020-11-27        | 2.1     | Sannareddy Ramesh | Modified data sets and added new problems |
| 2020-08-28        | 2.0     | Lavanya           | Moved lab to course repo in GitLab        |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
